torch/nn/modules/module.py

```diff
@@ -2131,20 +2146,18 @@ def state_dict(self, *args, destination=None, prefix="", keep_vars=False):
                 )
         for hook in self._state_dict_hooks.values():
             hook_result = hook(self, destination, prefix, local_metadata)
-            if hook_result is not None:
-                destination = hook_result
+            if not getattr(hook, "_from_public_api", False):
+                if hook_result is not None:
+                    destination = hook_result
```

Comment:
nit: should we have an else that aserts that hook_result is None for the new API?